src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunLibraryRootProvider.cs

```diff
@@ -131,15 +136,28 @@ private void RootMethods(TypeDesc type, string reason, IRootingServiceProvider r
                     // Skip them in library mode since they're not going to be callable.
                     continue;
                 }
+                catch (InvalidOperationException)
```

Comment:
Make a subclass of `TypeSystemException` instead of `InvalidOperationException` so we can collapse into a single catch above?

Reply:
Hmm, I'm afraid the problem is that this is what LayoutInt.AsInt throws for indeterminate types. We would need to ask @MichalStrehovsky whether it's fine to change the exception type thrown from that property getter.

Reply:
We catch `TypeSystemException` in several places already because we consider it non-fatal (junk user code/incomplete references that we can skip over). I don't think dereferencing indeterminately-sized type should be caught as a non-fatal error in those catch blocks.

What exactly triggers the exception? Can we just go over all types in the signature and check that their size is not indeterminate? `CheckCanGenerateMethod` already goes over all the parameters and return values.

Reply:
I think it's more or less what you say - the exception is implicitly thrown by calls to LayoutInt.AsInt from the ArgIterator. I guess the only advantage of reusing the GC ref map builder code is to avoid code duplication but I agree the relevant logic is probably simple enough to merit open-coding the checks, especially if CheckCanGenerateMethod already does this. I would just be worried if it turned out that there are tricky details w.r.t. handling return values in a register / on the stack / via an indirect pointer, things like that that are easy to mess up when reimplementing especially when they differ based on targeting architecture or ABI.

Reply:
> I would just be worried if it turned out that there are tricky details w.r.t. handling return values in a register / on the stack / via an indirect pointer, things like that that are easy to mess up when reimplementing especially when they differ based on targeting architecture or ABI.

If there's more to this than just checking if the parameters/return type have indeterminate size, is there anything that would guarantee these don't end up in the compilation graph anyway?

The way compilation works right now is that we start by compiling the roots that are established in this file. We would skip them here. So far so good. But then we keep compiling any other methods referenced from the roots, until we reach a fixpoint. Methods that have parameters of indeterminate size won't get added to the graph because the moment RyuJIT asks about those, we nip them in the bud. But if there's potentially more to it (not just the indeterminate size) as Tomas is alluding to, we might still end up seeing problematic methods in the graph (dragged in because some method we did have in the graph calls them). This needs to be understood. It is my belief that checking for indeterminate size is sufficient, but then again I haven't reviewed any of the GCRefMap code.

Reply:
I guess @trylek is probably referring to this:

https://github.com/dotnet/coreclr/blob/62bba3cc35b3ca761c542551f6140f054e88e682/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/GCRefMapBuilder.cs#L273

I looked at this code and didn't catch the `NotImplementedException`. I believe it is better for the compiler to crash early for now. So that we can determine if that actually happens during our testing. Without my change, it will be crashing much later.

Reply:
I was actually rather referring to this bit:

https://github.com/dotnet/coreclr/blob/5dc16c89311f4508ff05ffff15b4437bee3950a5/src/tools/crossgen2/Common/TypeSystem/Common/LayoutInt.cs#L50

NotImplementedException is one thing (missing implementation for some corner case of argument passing), the exception Simon started this thread on is InvalidOperationException which gets thrown in the property getter.

Reply:
Other than that, I basically agree with Michal's argumentation.